File: sydney/Kernel/Client/DataSource.cpp
// -*-Mode: C++; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// DataSource.cpp --
// 
// Copyright (c) 2002, 2004, 2005, 2008, 2009, 2023 Ricoh Company, Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 

namespace {
const char srcFile[] = __FILE__;
const char moduleName[] = "Client";
}

#include "SyDefault.h"
#include "SyReinterpretCast.h"
#include "Client/DataSource.h"
#include "Client/Connection.h"
#include "Client/Parameter.h"
#include "Client/Port.h"
#include "Client/Session.h"
#include "Client/PrepareStatement.h"

#include "Common/StringData.h"
#include "Common/Request.h"
#include "Common/Parameter.h"
#include "Common/IntegerData.h"
#include "Common/Message.h"

#include "Communication/LocalClientConnection.h"
#include "Communication/RemoteClientConnection.h"
#include "Communication/ConnectionSlaveID.h"

#include "Os/SysConf.h"
#include "Os/AutoCriticalSection.h"

#include "Exception/ConnectionRanOut.h"
#include "Exception/Unexpected.h"
#include "Exception/NotInitialized.h"
#include "Exception/BadArgument.h"

#include "ModAutoPointer.h"
#include "ModVector.h"
#include "ModOsDriver.h"

_TRMEISTER_USING
_TRMEISTER_CLIENT_USING

namespace
{
//
//	VARIABLE local
//	_$$::_iConnectionThreshold -- 新しいコネクションを作成する閾値
//
//	NOTES
//	新しいコネクションを作成する閾値。セッションの数が基準。デフォルト20
//
ParameterInteger
_iConnectionThreshold("Client_ConnectionThreshold", 20);

//
//	VARIABLE local
//	_$$::_iMaximumConnectionPoolCount -- ポートプールの最大値
//
//	NOTES
//	ポートプールの最大値。この値以上のポートがプールされた場合には余ったポートは削除される。
//
ParameterInteger
_iMaximumConnectionPoolCount("Client_MaximumConnectionPoolCount", 10);

//
//	VARIABLE local
//	_$$::_iCheckConnectionPoolPeriod -- ポートプールをチェックする間隔
//
//	NOTES
//	ポートプール数をチェックする間隔。
//	単位はミリ秒。ただし最小単位は500である。
//
ParameterInteger
_iCheckConnectionPoolPeriod("Client_CheckConnectionPoolPeriod", 60000);

}

//
//	FUNCTION public
//	Client::DataSource::DataSource -- コンストラクタ(InProcess用)
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
DataSource::DataSource()
	: m_iConnectionArgument(0),
	  m_iPortNumber(0),
	  m_iOpenCounter(0),
	  m_eProtocolVersion(Protocol::Version1)
{
	ModOsDriver::Memory::reset(m_pInstanceCount,
							   sizeof(int)*Client::Object::Type::ElementSize);
}

//
//	FUNCTION public
//	Client::DataSource::DataSource -- コンストラクタ(RemoteServer用)
//
//	NOTES
//
//	ARGUMENTS
//	const ModUnicodeString& cstrHostName_
//		ホスト名
//	int iPortNumber_
//		ポート番号
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
DataSource::DataSource(const ModUnicodeString& cstrHostName_,
					   int iPortNumber_)
	: m_iConnectionArgument(0),
	  m_cstrHostName(cstrHostName_),
	  m_iPortNumber(iPortNumber_),
	  m_iOpenCounter(0),
	  m_eProtocolVersion(Protocol::Version1)
{
	ModOsDriver::Memory::reset(m_pInstanceCount,
							   sizeof(int)*Client::Object::Type::ElementSize);
}

//
//	FUNCTION public
//	Client::DataSource::~DataSource -- デストラクタ
//
//	NOTES
//	デストラクタ。
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
DataSource::~DataSource()
{
	try
	{
		close();
	} catch (...) {}	// 例外は無視
}

//
//	FUNCTION public
//	Client::DataSource::open -- サーバと接続する
//
//	NOTES
//	
//	ARGUMENTS
//	Client::DataSource::Protocol::Value
//		プロトコルバージョン(default Protocol::Version1)
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::open(Protocol::Value eProtocolVersion_)
{
	Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);

	if (m_iOpenCounter++ == 0)
	{
		// プロトコルバージョンを設定する
		m_eProtocolVersion = eProtocolVersion_;

		ModAutoPointer<Port> pPort
			= getNewPort(Communication::ConnectionSlaveID::Any);

		pPort->open();

		//リクエストを送る
		const Common::Request request(Common::Request::BeginConnection);
		pPort->writeObject(&request);

		//クライアントのホスト名を送る
		const Common::StringData data(Os::SysConf::HostName::get());
		pPort->writeObject(&data);
		pPort->flush();

		//ステータスを得る
		pPort->readStatus();

		//クライアントコネクションを作成
		ModAutoPointer<Connection> pClientConnection
			= new Connection(*this, pPort.release());

		//配列に加える
		m_vecpClientConnection.pushBack(pClientConnection.release());

		//ポートチェックスレッドを起動する
		create();
	}
}

//
//	FUNCTION public
//	Client::DataSource::close -- サーバとの接続を切断する
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::close()
{
	Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);
		
	if (--m_iOpenCounter == 0)
	{
		//ポートチェックスレッドを停止する
		abort();
		join();

		{
			Os::AutoCriticalSection cAuto1(m_cPortMapLatch);

			//ポートプール内のものをすべて切断する
			PortMap::Iterator k = m_mapPort.begin();
			for (; k != m_mapPort.end(); ++k)
			{
				try
				{
					(*k).second->close();
				} catch (...) {}	// 例外は無視
				(*k).second->release();
			}
			m_mapPort.clear();
		}
		{
			//すべてのコネクションを切断する
			ModVector<Connection*>::Iterator i = m_vecpClientConnection.begin();
			for (; i != m_vecpClientConnection.end(); ++i)
			{
				(*i)->close();
				(*i)->release();
			}
			m_vecpClientConnection.clear();
			m_iConnectionArgument = 0;
		}
	}
}

//
//	FUNCTION public
//	Client::DataSource::release -- メモリーを解放する
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::release()
{
	delete this;
}

//
//	FUNCTION public
//	Client::DataSource::shutdown -- サーバを停止する
//
//	NOTES
//	
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::shutdown()
{
	Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);

	ModAutoPointer<Port> pPort
		= getNewPort(Communication::ConnectionSlaveID::Any);

	pPort->open();

	//リクエストを送る
	const Common::Request request(Common::Request::Shutdown);
	pPort->writeObject(&request);
	pPort->flush();

	//ステータスを得る
	pPort->readStatus();

	pPort->close();
}

//
//	FUNCTION public
//	Client::DataSource::createSession -- セッションを作成する
//
//	NOTES
//
//	ARGUMENTS
//	const ModUnicodeString& cstrDatabaseName_
//		データベース名
//
//	RETURN
//	Client::Session*
//		セッションオブジェクト
//
//	EXCEPTIONS
//
Session*
DataSource::createSession(const ModUnicodeString& cstrDatabaseName_)
{
	Port*	p = 0;

	{
		Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);

		Connection* pClientConnection = 0;

		try {
			//クライアントコネクションを得る
			pClientConnection = getClientConnection();
		} catch (Exception::NotInitialized&) {
			// close のみ実施されているので、open する
			try {
				open(m_eProtocolVersion);
				pClientConnection = getClientConnection();
			}
			catch (...) {}
		}

		if (pClientConnection == 0)
			_TRMEISTER_THROW0(Exception::ConnectionRanOut);

		//Workerを起動する
		int iWorkerID;
		try {
			p = pClientConnection->beginWorker(iWorkerID);
		} catch (Exception::ConnectionRanOut&) {
			if (getCount(Client::Object::Type::Session) != 0)
				throw;
			// サーバが再起動したかもしれないので、DataSourceを初期化する
			try {
				close();
				open(m_eProtocolVersion);
				pClientConnection = getClientConnection();
				p = pClientConnection->beginWorker(iWorkerID);
			} catch (...) {}
			if (p == 0)
				throw;
		}
	}
	
	ModAutoPointer<Port> pPort = p;

	// [<-] BeginSession
	const Common::Request request(Common::Request::BeginSession);
	pPort->writeObject(&request);

	// [<-] データベース名
	const Common::StringData data(cstrDatabaseName_);
	pPort->writeObject(&data);
	pPort->flush();

	// [->] セッション番号
	ModAutoPointer<Common::IntegerData> pSessionID = pPort->readIntegerData();

	// [->] ステータス
	pPort->readStatus();

	//セッションを確保
	Session* pSession = new Session(*this,
									cstrDatabaseName_,
									pSessionID->getValue());

	//コネクションを返す
	pushPort(pPort.release());

	//必要ならあたらしいクライアントコネクションを得る
	newClientConnect();

	return pSession;
}

//
//	FUNCTION public
//	Client::DataSource::createPrepareStatement -- 最適化結果を作成する
//
//	NOTES
//
//	ARGUMENTS
//	const ModUnicodeString& cstrDatabaseName_
//		データベース名
//	const ModUnicodeString& cstrStatement_
//
//	RETURN
//	Client::PrepareStatement*
//		最適化結果
//
//	EXCEPTIONS
//
PrepareStatement*
DataSource::createPrepareStatement(const ModUnicodeString& cstrDatabaseName_,
								   const ModUnicodeString& cstrStatement_)
{
	//クライアントコネクションを得る
	Connection* pClientConnection = getClientConnection();

	//Workerを起動する
	int iWorkerID;
	ModAutoPointer<Port> pPort = pClientConnection->beginWorker(iWorkerID);

	// [<-] BeginSession
	const Common::Request request(Common::Request::PrepareStatement);
	pPort->writeObject(&request);

	// [<-] データベース名
	const Common::StringData name(cstrDatabaseName_);
	pPort->writeObject(&name);

	// [<-] SQL文
	const Common::StringData stmt(cstrStatement_);
	pPort->writeObject(&stmt);
	pPort->flush();

	// [->] PrepareID
	ModAutoPointer<Common::IntegerData> pPrepareID = pPort->readIntegerData();

	// [->] ステータス
	pPort->readStatus();

	//最適化結果を確保
	PrepareStatement* pPrepareStatement
		= new PrepareStatement(*this, cstrDatabaseName_,
							   pPrepareID->getValue());

	//コネクションを返す
	pushPort(pPort.release());

	return pPrepareStatement;
}

//
//	FUNCTION public
//	Client::DataSource::isServerAvailable -- サーバの利用可能性を得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	bool
//		サーバが利用可能な場合はtrue、利用不可能な場合はfalse
//
//	EXCEPTIONS
//
bool
DataSource::isServerAvailable()
{
	// クライアントコネクションを得る
	Connection* pClientConnection = getClientConnection();
	// 利用可能性を問い合わせる
	return pClientConnection->isServerAvailable();
}

//
//	FUNCTION public
//	Client::DataSource::isDatabaseAvailable -- データベースの利用可能性を得る
//
//	NOTES
//
//	ARGUMENTS
//	Client::DataSource::Database::ID id_
//		データベースID
//		Database::Allが指定された場合はすべてのデータベースの利用可能性を
//		チェックする
//
//	RETURN
//	bool
//		データベースが利用可能な場合はtrue、利用不可能な場合はfalse
//
//	EXCEPTIONS
//
bool
DataSource::isDatabaseAvailable(Database::ID id_)
{
	// クライアントコネクションを得る
	Connection* pClientConnection = getClientConnection();
	// 利用可能性を問い合わせる
	return pClientConnection->isDatabaseAvailable(id_);
}

//
//	FUNCTION public
//	Client::DataSource::getClientConnection -- クライアントコネクションを得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Client::Connection*
//		クライアントコネクション
//
//	EXCEPTIONS
//
Connection*
DataSource::getClientConnection()
{
	Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);

	if (m_vecpClientConnection.getSize() == 0)
		// すでにクローズされている
		_TRMEISTER_THROW0(Exception::NotInitialized);

	if (m_iConnectionArgument
		>= static_cast<int>(m_vecpClientConnection.getSize()))
		m_iConnectionArgument = 0;

	return m_vecpClientConnection[m_iConnectionArgument++];
}

//
//	FUNCTION public
//	Client::DataSource::popPort -- ポートを削除する
//
//	NOTES
//	ポートを削除する
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Client::Port*
//		プールされていない場合は0を返す
//
//	EXCEPTIONS
//
Port*
DataSource::popPort()
{
	Os::AutoCriticalSection cAuto(m_cPortMapLatch);

	Port* pPort = 0;

	PortMap::Iterator i = m_mapPort.begin();
	if (i != m_mapPort.end())
	{
		pPort = (*i).second;
		m_mapPort.erase(i);
	}

	return pPort;
}

//
//	FUNCTION public
//	Client::DataSource::pushPort -- ポートを追加する
//
//	NOTES
//	ポートを追加する
//
//	ARGUMENTS
//	Clinet::Port* pPort_
//		コネクション
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::pushPort(Port* pPort_)
{
	Os::AutoCriticalSection cAuto(m_cPortMapLatch);
	m_mapPort.insert(pPort_->getSlaveID(), pPort_);
}

//
//	FUNCTION public
//	Client::DataSource::getNewPort -- 新しいポートを得る
//
//	NOTES
//
//	ARGUMENTS
//	int iSlaveID_
//		スレーブID
//
//	RETURN
//	Client::Port* 
//		 新しいポートのインスタンス
//
//	EXCEPTIONS
//
Client::Port*
DataSource::getNewPort(int iSlaveID_)
{
	Port* pPort;

	if (m_iPortNumber == 0)
	{
		//Local
		pPort = new Port(m_eProtocolVersion, iSlaveID_);
	}
	else
	{
		//RemoteServer
		pPort = new Port(m_cstrHostName, m_iPortNumber,
						 m_eProtocolVersion, iSlaveID_);
	}

	return pPort;
}

//
//	FUNCTION public
//	Client::DataSource::incrementCount -- インスタンス数をインクリメントする
//
//	NOTES
//
//	ARGUMENTS
//	Client::Object::Type::Value eType_
//		クラス種別
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
void
DataSource::incrementCount(Client::Object::Type::Value eType_)
{
	Os::AutoCriticalSection cAuto(m_cInstanceLatch);
	m_pInstanceCount[eType_]++;
}

//
//	FUNCTION public
//	Client::DataSource::decrementCount -- インスタンス数をデクリメントする
//
//	NOTES
//
//	ARGUMENTS
//	Client::Object::Type::Value eType_
//		クラス種別
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
void
DataSource::decrementCount(Client::Object::Type::Value eType_)
{
	Os::AutoCriticalSection cAuto(m_cInstanceLatch);
	m_pInstanceCount[eType_]--;
}

//
//	FUNCTION public
//	Client::DataSource::getCount -- インスタンス数を得る
//
//	NOTES
//
//	ARGUMENTS
//	Client::Object::Type::Value eType_
//		クラス種別
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
int
DataSource::getCount(Client::Object::Type::Value eType_)
{
	Os::AutoCriticalSection cAuto(m_cInstanceLatch);
	return m_pInstanceCount[eType_];
}

//
//	FUNCTION public static
//	Client::DataSource::createDataSource -- データソースを得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Client::DataSource*
//		新しいデータソース
//
//	EXCEPTIONS
//
DataSource*
DataSource::createDataSource()
{
	return new DataSource;
}

//
//	FUNCTION public static
//	Client::DataSource::createDataSource -- データソースを得る
//
//	NOTES
//
//	ARGUMENTS
//	const ModUnicodeString& cstrHostName_
//		ホスト名
//	int iPortNumber_
//		ポート番号
//
//	RETURN
//	Client::DataSource*
//		新しいデータソース
//
//	EXCEPTIONS
//
DataSource*
DataSource::createDataSource(const ModUnicodeString& cstrHostName_,
							 int iPortNumber_)
{
	//引数チェック
	if (cstrHostName_.getLength() == 0 || iPortNumber_ <= 0)
	{
		_TRMEISTER_THROW0(Exception::BadArgument);
	}
	return new DataSource(cstrHostName_, iPortNumber_);
}

//
//	FUNCTION private
//	Client::DataSource::runnable -- ポートプールを管理するスレッド
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::runnable()
{
	//チェック間隔の単位
	const int TIME_UNIT = 500;

	while (true)
	{
		int iCount = 0;
		int iTotalCount = _iCheckConnectionPoolPeriod.get() / TIME_UNIT;

		while (true)
		{
			ModOsDriver::Thread::sleep(TIME_UNIT);
			if (isAborted() == true)
				//終了
				return;
			iCount++;
			if (iCount >= iTotalCount)
				break;
		}

		ModVector<int> veciSlaveID;

		{
			// ポートプールの数をチェックする
			Os::AutoCriticalSection cAuto(m_cPortMapLatch);

			if (static_cast<int>(m_mapPort.getSize())
				> _iMaximumConnectionPoolCount.get())
			{
				//超えているので超えている分を削除する
				int s = m_mapPort.getSize()
					- _iMaximumConnectionPoolCount.get();
				PortMap::Iterator e = m_mapPort.begin();
				for (int i = 0; i < s; ++i)
				{
					Port* pPort = (*e).second;
					veciSlaveID.pushBack(pPort->getSlaveID());
					pPort->close();
					pPort->release();
					e++;
				}

				//マップから削除
				m_mapPort.erase(m_mapPort.begin(), e);
			}
		}

		if (veciSlaveID.getSize())
		{
			//削除したものがあるので、サーバに知らせる

			//クライアントコネクションを得る
			Connection* pClientConnection = getClientConnection();
			//最適化結果を削除する
			pClientConnection->disconnectPort(veciSlaveID);
		}
	}
}

//
//	FUNCTION private
//	Client::DataSource::newClientConnect
//		-- 新しいクライアントコネクションを確保する
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//
void
DataSource::newClientConnect()
{
	Os::AutoCriticalSection cAuto(m_cClientConnectionLatch);

	if (getCount(Client::Object::Type::Session)
		> getCount(Client::Object::Type::Connection) *
		_iConnectionThreshold.get())
	{
		//セッション数が超えた -> 新しいコネクションをはる

		//クライアントコネクションを得る
		Connection* pClientConnection = getClientConnection();

		//新しいクライアントコネクションを得る
		Connection* pNewClientConnection = pClientConnection->beginConnection();

		//配列に加える
		m_vecpClientConnection.pushBack(pNewClientConnection);
	}
}

//
//	Copyright (c) 2002, 2004, 2005, 2008, 2009, 2023 Ricoh Company, Ltd.
//	All rights reserved.
//

File: sydney/Python/src/doquedb/client/datasource.py
# Copyright (c) 2024 Ricoh Company, Ltd.
#
# Licensed under the Apache License, Version 2.0 (the License);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
datasource.py -- クライアントのデータソース
"""
# TODO:
#    port_map, session_mapなどは外部で管理できるような実装に変える
#    :func: `new_client_connection`の実装変更
from typing import Optional, List, Dict

import socket

from ..exception import exceptions
from .object import Object
from .constants import ProtocolVersion
from .connection import Connection
from .session import Session
from .port import Port
from ..common.serialdata import Request
from ..common.scalardata import IntegerData, StringData
from ..port.constants import ConnectionSlaveID
from ..port.authorizemode import AuthorizeMode


class DataSource(Object):
    """データソースクラス.

    クライアント側のデータソースをあらわすクラス. 接続に必要な情報を保持する.

    Args:
        hostname (str): 接続先のホスト名
        portnum (int): ポート番号

    Attributes:
        __connection_list (list): コネクションのリスト
        __connection_element (int): 最近返したコネクションの要素番号
        __portmap (dict): ポートのハッシュマップ(辞書)
        __expunge_port (list): クライアントが廃棄したポート

        __hostname (str): 接続先のホスト名
        __portnumber (int): ポート番号
        __protocol_ver (int): プロトコルバージョン

        __sessionmap (dict): セッションのハッシュマップ(辞書)
        __is_closed (bool): クローズしたかどうか
        __masterID (int): マスターID(=プロトコルバージョン)
        __authorization (int): 認証方式
    """
    # 全データベースをあらわす ID
    DATABASE_ALL = 0xffffffff

    # １つのコネクションが管理する最大セッション数 (JDBCからの値)
    CONNECTION_THRESHOLD = 20
    # ポートプールチェック間隔の最小単位(ms) (JDBCからの値)
    TIME_UNIT = 500
    # ポートプールの最大ポート数 (JDBCからの値)
    MAXIMUM_CONNECTION_POOL_COUNT = 100
    # ポートプール数をチェックする間隔(ms) (JDBCからの値)
    CHECK_CONNECTION_POOL_PERIOD = 60000

    def __init__(self, hostname: str, portnum: int) -> None:
        super().__init__(Object.DATA_SOURCE)
        # メンバー変数の初期化
        self.__connection_list: List[Connection] = []
        self.__connection_element = 0
        self.__portmap: Dict[int, Port] = {}
        self.__expunge_port: List[int] = []
        self.__hostname = hostname
        self.__portnum = portnum
        self.__protocol_ver = 0
        self.__session_map: Dict[int, Session] = {}
        self.__is_closed = False
        self.__master_id = 0
        self.__authorization = 0

    @property
    def master_id(self) -> int:
        """マスターIDのゲッター"""
        return self.__master_id

    @property
    def authorization(self) -> int:
        """認証方式のゲッター."""
        return self.__authorization

    def open(self, protocol_ver: int) -> None:
        """データソースをオープンする.

        Args:
            protocol_ver(int): プロトコルバージョン
        """
        self.__protocol_ver = protocol_ver

        # 認証方式が指定されていない場合はパスワード認証を指定する
        if (self.__protocol_ver & AuthorizeMode.MaskMode.value) \
                == AuthorizeMode.NONE.value:
            self.__protocol_ver |= AuthorizeMode.PASSWORD.value

        # 新たにポートを得る
        port = self.new_port(ConnectionSlaveID.ANY.value)
        port.open()

        # [<-] リクエスト
        port.write_object(Request(Request.request_map['BEGIN_CONNECTION']))
        # [<-] ホスト名
        localhost = socket.gethostbyaddr(
            socket.gethostbyname(socket.gethostname()))[0]
        port.write_object(StringData(localhost))
        port.flush()
        # [->] ステータス
        port.read_status()
        # クライアントコネクションを作成する
        client_connection = Connection(self, port)
        # 配列に加える
        self.__connection_list.append(client_connection)
        # マスターIDを保存する
        self.__master_id = port.master_id
        # 認証方式を保存する
        self.__authorization = port.authorization

        self.__is_closed = False

    def create_session(self,
                       dbname: str,
                       username: Optional[str] = None,
                       password: Optional[str] = None
                       ) -> Session:
        """新しくセッションを作成する.

        Args:
            dbname(str): データベース名
            username(Optional[str]): ユーザー名
            password(Optional[str]): パスワード

        Raises:
            UnexpectedError: クライアントコネクションが取得できなかった場合

        TODO:
            セッション開始の例外処理の詳細実装
        """
        # プロトコルバージョン4以下はユーザー管理非対応
        if self.__master_id < ProtocolVersion.PROTOCOL_VERSION5.value:
            username = None
            password = None

        port: Optional[Port] = None

        # クライアントコネクションを得る.
        client_connection = self.get_client_connection()

        # ワーカの起動
        try:
            if client_connection is None:
                # 一旦クローズされているので再度 open する
                self.open(self.__protocol_ver)
                client_connection = self.get_client_connection()
            # ワーカを起動する
            if client_connection:
                port = client_connection.begin_worker()
            else:
                raise exceptions.UnexpectedError(
                    'failed to get client connection (1)')
        except Exception as err:
            if (self.session_exist()):
                # まだ利用中のセッションがあるので再接続しない
                raise err

            # サーバが再起動したかもしれないので、データソースを初期化する
            try:
                self.close()
                self.open(self.__protocol_ver)
                client_connection = self.get_client_connection()
                if client_connection:
                    port = client_connection.begin_worker()
                else:
                    raise exceptions.UnexpectedError(
                        'failed to get client connection (2)')
            except Exception as err:
                if port is None:
                    # 再初期化してもダメ
                    raise err

        assert port is not None
        session_id: Optional[IntegerData] = None

        # 引数にusernameとpasswordがあった場合
        if username and password:
            # セッションの開始
            try:
                # [<-] BeginSession
                port.write_object(
                    Request(Request.request_map['BEGIN_SESSION2']))
                # [<-] データベース名
                port.write_object(StringData(dbname))
                # [<-] ユーザー名
                port.write_object(StringData(username))
                # [<-] パスワード
                port.write_object(StringData(password))
                port.flush()

                # [->] セッション番号
                session_id = port.read_integerdata()
                # [->] ステータス
                port.read_status()
            except exceptions.DatabaseError as dbe:
                if port.is_reuse:
                    self.push_port(port)
                else:
                    port.close()
                raise dbe
            except exceptions.UnexpectedError as ue:
                port.close()
                raise ue
            except Exception as err:
                port.close()
                raise err

        # 引数がdbnameのみの場合
        else:
            assert username is None and password is None
            try:
                # [<-] BeginSession
                port.write_object(
                    Request(Request.request_map['BEGIN_SESSION']))
                # [<-] データベース名
                port.write_object(StringData(dbname))
                port.flush()

                # [->] セッション番号
                session_id = port.read_integerdata()
                # [->] ステータス
                port.read_status()
            except exceptions.DatabaseError as dbe:
                if port.is_reuse:
                    self.push_port(port)
                else:
                    port.close()
                raise dbe
            except exceptions.UnexpectedError as ue:
                port.close()
                raise ue
            except Exception as err:
                port.close()
                raise err

        # コネクションをプールする
        self.push_port(port)

        # セッションを新しく作成して登録する.
        session = Session(self, dbname, session_id.value, username)
        self.__session_map[session.id_] = session

        # 必要なら新しいクライアントコネクションを得る
        self.new_client_connection()

        return session

    def get_client_connection(self) -> Optional[Connection]:
        """コネクション管理クラスを得る.

        ラウンドロビン方式でコネクション管理クラスを入手する.

        Returns:
            Connection: コネクション管理クラス
        """
        if len(self.__connection_list) == 0:
            return None
        if self.__connection_element >= len(self.__connection_list):
            self.__connection_element = 0

        connection = self.__connection_list[self.__connection_element]
        self.__connection_element += 1
        return connection

    def new_client_connection(self) -> None:
        """セッション数がCONNECTION_THRESHOLDを越えた場合に、
        新しいクライアントコネクションを作成する.

        TODO:
            connection_listは今の実装では増え続けてしまうので、
            定期的に監視するように実装を変える必要がある。
        """
        size = len(self.__session_map)
        # CONNECTION_THRESHOLDをセッション数が越えたので、
        # 新しいクライアントコネクションを作る.
        if size >= self.CONNECTION_THRESHOLD * len(self.__connection_list):
            client_connection = self.get_client_connection()
            # 新しいクライアントコネクションを得る
            assert client_connection is not None
            new_client_connection = client_connection.begin_connection()
            # 配列に加える
            self.__connection_list.append(new_client_connection)

    def pop_port(self) -> Optional[Port]:
        """ポートプールからポートを取り出す.

        Returns:
            Port: ポートプールから受け取ったポート
        """
        port = None
        if len(self.__portmap) > 0:
            k, v = self.__portmap.popitem()
            port = v
        return port

    def push_port(self, port: Port) -> None:
        """ポートプールにポートを挿入する.

        Args:
            port_(Port): ポート
        """
        port.reset()
        self.__portmap[port.slave_id] = port

    def new_port(self, slave_id: int) -> Port:
        """新しいポートのインスタンスを得る

        Args:
            slave_id(int): スレーブID

        Returns:
            Port: 新しいポートインスタンス
        """
        return Port(
            self.__hostname, self.__portnum, self.__protocol_ver, slave_id)

    def expunge_port(self, port: Port) -> None:
        """廃棄したポートを登録する.

        Args:
            port (Port): 廃棄するポート
        """
        self.__expunge_port.append(port.slave_id)
        port.close()

    def session_exist(self) -> bool:
        """セッションが存在しているかどうか

        Returns:
            bool: 存在している場合True, それ以外はFalse
        """
        if len(self.__session_map) == 0:
            return False
        return True

    def remove_session(self, id: int) -> None:
        """セッションを削除する.
        """
        del self.__session_map[id]

    def is_server_available(self) -> bool:
        """サーバの利用可能性を得る.

        Returns:
            bool: サーバが利用可能な場合はTrue, 利用不可能な場合はFalse
        """
        # クライアントコネクションを得る
        client_connection = self.get_client_connection()
        assert client_connection is not None
        # 利用可能性を問い合わせる
        return client_connection.is_serever_available()

    def is_database_available(self) -> bool:
        """任意のデータベースの利用可能性を得る.

        Returns:
            bool: データベースが利用可能な場合はTrue , 利用不可能な場合はFalse
        """
        pass

    def shutdown(self,
                 username: Optional[str] = None,
                 password: Optional[str] = None
                 ) -> None:
        """サーバを停止する.
        """
        # 新たにポートを得る
        port = self.new_port(ConnectionSlaveID.ANY.value)
        port.open()

        old = False
        # ユーザ認証のあるサーバに対しての処理
        if username and password:
            try:
                # [<-] リクエスト
                port.write_object(Request(Request.request_map['SHUTDOWN2']))
                port.write_object(StringData(username))
                port.write_object(StringData(password))
                port.flush()
                # [->] ステータス
                port.read_status()
            except exceptions.DatabaseError:
                old = True
            finally:
                port.close()
        # 通常の処理
        else:
            # [<-] リクエスト
            port.write_object(Request(Request.request_map['SHUTDOWN']))
            port.flush()
            # [->] ステータス
            port.read_status()

            port.close()

        # 古いサーバと接続しているので、古い shutdown 処理を実行
        if old is True:
            self.shutdown()

    def close(self) -> None:
        """データソースをクローズする.
        """
        if self.__is_closed:
            return

        # 全てのセッションをクローズする
        for session in self.__session_map.values():
            session.close_internal()
        self.__session_map.clear()

        # 全てのコネクションをクローズする
        for i in range(len(self.__connection_list)):
            c = self.__connection_list[i]
            c.close()
        self.__connection_list.clear()

        # 全てのポートをクローズする
        for port in self.__portmap.values():
            port.close()
        self.__portmap.clear()
        self.__expunge_port.clear()

        self.__is_closed = True
